components/style/traversal.rs

```diff
@@ -284,6 +301,14 @@ pub trait DomTraversal<E: TElement> : Sync {
 
     /// Creates a thread-local context.
     fn create_thread_local_context(&self) -> Self::ThreadLocalContext;
+
+    /// Whether we're performing a parallel traversal.
+    ///
+    /// NB: We do this check on runtime. We could guarantee correctness in this
+    /// regard via the type system via a `TraversalKind` trait for this trait,
+    /// that could be one of two concrete types. It's not clear whether the
+    /// potential code size impact of that is worth it.
+    fn kind(&self) -> TraversalKind;
```

Comment:
Yeah I agree that doing this dynamically makes the most sense.

I think the ergonomics would be a bit better if we just had is_parallel() here, rather than implementing it on TraversalKind, then exposing kind(), then having the caller check that. We can always expose more information as-needed if we ever develop new traversal variants.